threadpool from nifty/vigra in xtl

DerThorsten opened the pull request:
threadpool from nifty/vigra in xtl. What do you think about such a threadpool somewhere in your Stack?

The threadpool itself
``` c++
#include <xtl/xthreadpool.hpp>

auto n_workers = 16;
xtl::xthreadpool pool(n_workers);

// add function to pool
auto fut = pool.enqueue([&](int thread_index){
     longRunningStuff();
});


// add more function  to pool
...
// get result 
auto res  = fut.get();

// wait for completion
pool.wait();
```

integer based parallel foreach
``` c++
#include <vector>
#include <xtl/xparallel_foreach.hpp>

auto vec = std::vector<int>(10000);
auto n_workers = 16;
xtl::xparallel_foreach(vec.size(), n_workers, [&](auto thread_index, auto i){
    vec[i] = someLongRunningThing(i);
});
```
Container /iterator based foreach 
``` c++
#include <vector>
#include <xtl/xparallel_foreach.hpp>

struct Foo{
    int val;
};
const auto n_workers = 8;
const auto size = 100;
std::vector<Foo> vec(size);  

// use container based api
xtl::xparallel_foreach(vec, n_workers, [&](auto thread_index, Foo & foo){
  foo.val = 1;
});

// use iterator based api
xtl::xparallel_foreach(vec.begin(),  vec.end(),  n_workers, [&](auto thread_index, Foo & foo){
  foo.val = 1;
});
```
A few overloads are still missing specially to pass an already created threadpool to parallel_foreach.
The implementation of the input_iterator overload might still  suboptimal and has still a few todos.

JohanMabille commented on include/xtl/xthreadpool.hpp:
```diff
@@ -0,0 +1,286 @@
+        to zero (i.e. synchronous execution), regardless of the value of \arg n. This
+        is useful for debugging.
+     */
+    xthreadpool(const int n)
```
`const` is unecessary here; the number of threads should be an unsigned integer.

DerThorsten replied:
I know that it is unnecessary from a technical point of view, but I personally like the const.
Seems like I am not the only one https://stackoverflow.com/questions/117293/use-of-const-for-function-parameters (of course not wanting the const has also big support).
But if you prefer to not have them I can get ride of them

JohanMabille commented on include/xtl/xthreadpool.hpp:
```diff
@@ -0,0 +1,286 @@
+    // helper function to init the thread pool
+    void init();
+
+    xparallel_options options_;
```
Not sure that we need to keep the `xparallel_options` as a class member since the number of threads is available through `workers.size()`.

JohanMabille commented on include/xtl/xthreadpool.hpp:
```diff
@@ -0,0 +1,286 @@
+
+    size_t n_threads() const
+    {
+        return workers.size() == 0 ? std::size_t(1) : workers.size();
```
I probably missed the point of this since I find it confusing. This returns the same value whether I create a thread pool with a single thread or an empty thread pool. In the former case the total number of threads is 2 (main thread + the one created in the pool), in the latter case only 1.

DerThorsten replied:
One of the main features of this threadpool is the thread_index / worker_index.
All lambdas get this index as an argument. The idea is simple: often one needs per-thread-data structures.
Consider the case when no extra thread was requested (we want to run in the main thread).
The thread_index / worker_index is 0 in this case (it would also be 0 with just a asingle worker).
``` c++
struct PerThreadData{
};
auto n_threads = 0;
xthreadpool pool(n_threads);
std::vector<PerThreadData> vec(pool.n_threads());

pool.enqueue([&](auto worker_index){
   // get per thread data
   auto & data = vec[worker_index];
});

```
if `pool.n_threads()`  needs to return something non-negative.
But I somehow agree that this might be too confusing. 
Do you have a better Idea how to call this function?
Or one needs to write:
``` c++
std::vector<PerThreadData> vec(pool.n_threads() == 0 ? 1 : pool.n_threads();

```

JohanMabille commented on include/xtl/xthreadpool.hpp:
```diff
@@ -0,0 +1,286 @@
+    int n_threads() const
+    {
+        return std::max(1,numThreads_);
+    }
```
See my comment on the same method in the thread pool.

JohanMabille commented on include/xtl/xparallel_foreach.hpp:
```diff
@@ -0,0 +1,349 @@
+            const INTEGER size,
+            xthreadpool & pool,
+            F && f
+        ){
```
Formatting:
- upper case names are reserved for macros
- no space between a type name and '&' or '&&'

I don't understand why you use SFINAE here since the method is not overloaded / specialized elsewhere.

JohanMabille commented:
Hi,

Thanks for implementing this. Indeed we need a thread pool so we can make `xtensor` assignment multithreaded without the extra cost of creating the thread in each assignment loop.

See my comments in the review below (essentially on the thread pool for now, I'll go through the `for_each` implementations once the thread pool is fixed).

DerThorsten commented:
I removed the options class and replaced it with a class enum, but atm I am unhappy with the name `n_thread_settings`. Feel free to suggest a better name. 
I'll try to do the rest of the desired changes today or  tomorrow.

Greetings

wolfv commented on include/xtl/xparallel_foreach.hpp:
```diff
@@ -68,9 +68,9 @@ namespace xtl
         ){
             auto workload = size;
-            const auto work_per_thread = pool.n_worker_threads();
+            const auto work_per_thread = workload /  pool.n_worker_threads();
```
let's also be consistent with spaces around operators. **always** 1 space (`workload / pool.n...`).

Also, above is `double(work_per_thread)/double(XTL_DEFAULT_CHUNKING_FACTOR)` which should have spaces etc :)

DerThorsten commented:
I'll rewrite / cleanup some parts of parallel_foreach  tomorrow, so maybe wait with reviewing

wolfv commented:
awesome work!

DerThorsten commented:
It should be noted that the threadpool is very much based on:
https://github.com/progschj/ThreadPool with the crucial difference that a  every enqueued lambda 
get a consecutive worker_index.
With that it is very easy to implement a algorithms where each thread needs to store its own data.

wolfv commented:
@DerThorsten Why did you close the PR?

wolfv commented:
btw a good command is `git rebase --interactive origin/master` to select/deselect unnecessary commits :)

DerThorsten commented:
I did not want to many notifications for you guys since i'll commit more

JohanMabille commented:
@DerThorsten you can squash commits to reduce their number, and then force push.

wolfv commented:
Closing for lack of activity. 

We've decided to go with TBB in xtensor :) Hope that's all fine for you, Thorsten.
